module/retrymiddleware/after_consecutive_failures.go

```diff
@@ -0,0 +1,48 @@
+package retrymiddleware
+
+import (
+	"github.com/sethvargo/go-retry"
+	"sync"
+	"time"
+)
+
+// OnMaxConsecutiveFailures func that gets invoked when max consecutive failures has been reached
+type OnMaxConsecutiveFailures func(totalAttempts, clientIndex int)
+
+// AfterConsecutiveFailures returns a retry middleware that keeps track of the current number of attempts as well as the current client index.
+// When maxConsecutiveFailures has been reached the client index will be incremented or reset to 0 and the middleware will invoke f with the new
+// client index.
+func AfterConsecutiveFailures(maxConsecutiveFailures, numOfClients int, next retry.Backoff, f OnMaxConsecutiveFailures) retry.Backoff {
+	var l sync.Mutex
+
+	var attempt, client int
```

Comment:
This resets the client index to 0 each time we submit a new request. Suppose we have clients [a,b,c,d], and a and b are unavailable. Then we'll need to wait for failures on a and b each time before we can successfully send a request to c.

Reply:
Good call out, better to take into account **_last successful client_**

Reply:
Now that i'm thinking about it more, if we have to reset client index to 0 or we will be breaking the guarantee of client priority. We would also probably want those failure calls for metrics and logging.

Reply:
I think we should prioritize picking a client that is most likely to result in a successful request as quickly as possible, which in my mind would be that last client which was successful in a previous request.

Reply:
https://github.com/onflow/flow-go/pull/1556/commits/41cf3995759685b89b6b25c509fa5b02fefc0555